lib/api/CDataFrameAnalyzer.cc

```diff
@@ -53,54 +62,90 @@ bool CDataFrameAnalyzer::handleRecord(const TStrVec& fieldNames, const TStrVec&
     // Note if the the control message field name is missing the analysis must
     // be triggered to run by calling run explicitly.
 
-    bool result{true};
-    if (this->readyToReceiveControlMessages() == false) {
-        result = this->prepareToReceiveControlMessages(fieldNames);
+    if (m_AnalysisSpecification == nullptr || m_AnalysisSpecification->bad()) {
+        LOG_TRACE(<< "Specification is bad");
+        return false;
+    }
+
+    if (this->readyToReceiveControlMessages() == false &&
+        this->prepareToReceiveControlMessages(fieldNames) == false) {
+        return false;
+    }
+
+    if (this->sufficientFieldValues(fieldValues) == false) {
+        LOG_TRACE(<< "Expected " << m_AnalysisSpecification->numberColumns()
```

Comment:
I agree these should be recorded at error level somehow. However, the question is do we want to generate one error for every single row or a summary.

This is wrapped up with what we're planning to do for monitoring, but the consensus last time we discussed was we should maintain counters for problems which could occur many times.

Reply:
apart from the log level, i think it is interesting how we transport the message back to the user, logs are overwhelming and most people do not read them until they are aware of some big trouble. So somehow the job status should indicate a problem/error and show a message (doing it like forecast is an option, but it could also be something else)

Reply:
I agree. Some messages definitely need to be displayed more visibly to the user. However, IMO some messages are going to remain very difficult for a user to act on and I think these appearing in logs is appropriate. Remember as well we have two use cases: interactive and realtime. In real time we're going to need a different mechanism for surfacing important errors. Let's continue to work on the design for monitoring offline.